chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,142 @@
+7. **A:** The role of the NameNode is to save metadata about the file and to manage where they are saved, in addition to managing and saving the FS structure and also replicate / delete replications as needed.
+the metadata is stored in memory and also in a persistant FSImage file, in addition to a local editlog which includes the logs of all the eddits that have happened to the file system structure.
+the namenodes also monitors all the data nodes and knows which are alive and which are down.
+
 8. **Q:** What is a DataNode in HDFS?
+8. **A:** A DataNode is the node that saves the data itself in the form of blocks with a max size that was configured beforehand in the hdfs-site.
+The datanode also performs the computations and analysis of the data on the machines which allows for less movement of data.
+the data nodes send a block report every 6 hours to the namenode which includes all the blocks stored in this data node, in addition to the name of the file, the path ("virtual path") and the block count.
+the data node stores metadata about the file as well, with the same name of the file (blk_{GeneratedNumber}) with the suffix ".meta" which includes the file name, the file path, the block count, the block id, permissions of the file, owner, updated at and other time metadata and file size.
+
 9. **Q:** How does HDFS achieve fault tolerance?
+9. **A:** Hdfs achieves fault tolorence by using *redundency and replication* - duplicating the data multiple times and storing it on different machines so that if hardware failure were to occur, other machines store the replicas so they can be read from there, *rack awareness* - hdfs knows the physical network topology of the datanodes and namenodes and it uses it to better spread the replicas so that even if a rack falls or a network failure occurs, there will still be available replicas on a different switch/rack (if 3 replicas are set, one replica will be stored in the same place as the namenode and the 2 others will be stored in a different rack), in addition to *journal nodes* - nodes that store edit logs of the file system which can be used to reconstruct the file system structure in case of namenode failure and *secondary namenode* - a namenode that takes the edit logs file from the main namenode and constructs the fs image using the current fsimage and updates it in the main namenode, the reason for this is that usually the fs image only gets updated when the namenode restarts which can take a while which will cause the name node to fail the readiness check or crash, so the secondary namenode was created to mitigate this issue by periodically reconstructing the fs image and clearing the edit logs and *Standby namenode* - which waits and periodically syncs with the main name node and is there in case the main namenode crashes, the secondary namenode starts getting the requests and monitoring the dœ†a nodes.
+
 10. **Q:** Can you explain rack awareness in HDFS?
```

Comment:
Nice answer! but rack awarness isn't dependant only on the physical network topology. Please find out if I can configure a "logical" rack awarness